<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />


<meta name="date" content="2024-07-29" />

<title>SpatialDeX: A Reference-Free Method for Cell Type Deconvolution of Spatial Transcriptomics Data in Solid Tumors</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SpatialDeX: A Reference-Free Method for
Cell Type Deconvolution of Spatial Transcriptomics Data in Solid
Tumors</h1>
<h4 class="author">Xinyi Liu<sup>1</sup>, Gongyu Tang<sup>1,3</sup>,
Yuhao Chen<sup>1</sup>, Yuanxiang Li<sup>1</sup>, Hua Li<sup>4,5</sup>,
and Xiaowei Wang<sup>1,2</sup></h4>
<address class="author_afil">
<sup>1</sup>Department of Pharmacology and Regenerative Medicine,
University of Illinois at Chicago, <sup>2</sup>University of Illinois
Cancer Center, Chicago, IL, USA; <sup>3</sup>Department of Mechanical
Engineering and Materials Science, <sup>4</sup>Department of Radiation
Oncology, Washington University in St. Louis, St. Louis, MO, USA;
<sup>5</sup>Department of Bioengineering, University of Illinois
Urbana-Champaign, Urbana, IL.<br><h4 class="date">2024-07-29</h4>


<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction">1
Introduction</a></li>
<li><a href="#installation" id="toc-installation">2 Installation</a>
<ul>
<li><a href="#prerequisites" id="toc-prerequisites">2.1
Prerequisites</a></li>
<li><a href="#installing" id="toc-installing">2.2 Installing</a></li>
</ul></li>
<li><a href="#running-spatialdex" id="toc-running-spatialdex">3 Running
SpatialDeX</a>
<ul>
<li><a href="#running-deconvolution-example" id="toc-running-deconvolution-example">3.1 Running deconvolution
example</a></li>
<li><a href="#get-the-weighted-feature-coefficient" id="toc-get-the-weighted-feature-coefficient">3.2 Get the weighted
feature coefficient</a></li>
</ul></li>
<li><a href="#sessioninfo" id="toc-sessioninfo">sessionInfo</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1>1 Introduction</h1>
<p>SpatialDeX, a novel reference-free deconvolution method to estimate
the proportion of cell types at the spot level in spatial
transcriptomics (ST) data. Our hypothesis is that a complex mixture of
various cell types within a spot can be estimated by referencing
individual cell type signatures. Thus, SpatialDex offers a built-in
feature coefficient matrix to estimate cell type proportions in each
spot, eliminating the need for additional reference data. To this end,
we developed a linear regression model to compute the feature
coefficient matrix.</p>
<p>Let A represent the M × N feature matrix of M spots and N features. B
is the M × K binary cell identity matrix, where bjX = 1 if spot j
contains cell type X (j = 1, 2, …, M; X = 1, 2, …, K), and bjX = 0
otherwise. The relationship between cell types and features is indicated
in a latent coefficient matrix W with size K × N. The feature matrix A
is modeled as follows:</p>
<p><span class="math display">\[
A = BW + \epsilon
\]</span> where <span class="math inline">\(\epsilon\)</span> is a noise
term following a Gaussian distribution with a mean of zero. The
coefficient matrix W can be estimated through ridge regression:</p>
<p><span class="math display">\[
W = (B^T B + \lambda I)^{-1} B^T A
\]</span></p>
<p>Where <span class="math inline">\(I\)</span> is the identity matrix
and <span class="math inline">\(\lambda\)</span> was determined through
cross-validation using the training dataset. To compute the p values for
each feature, we conducted a permutation test by randomly assigning cell
type identities to the training spots 10,000 times. The FDR was then
calculated using the Benjamini-Hochberg method. Only features with the
coefficient &gt; 0 and FDR &lt; 0.001 were retained. The filtered <span class="math inline">\(W\)</span> was then used to generate <span class="math inline">\(B&#39;\)</span>, an estimated cell type score
matrix for new ST testing data as follows:</p>
<p><span class="math display">\[
B&#39; = A&#39;W(W^T W + \lambda&#39;I)^{-1}
\]</span></p>
<p>Where <span class="math inline">\(\lambda&#39;\)</span> was
determined through cross-validation and <span class="math inline">\(A&#39;\)</span> represents the feature matrix of
new ST testing data.</p>
<p>After obtaining estimated cell type score matrix through the linear
regression model, we first replaced the negative scores with zero, and
then converted the scores for each spot into percentage values. For each
spot, if the largest percentage was below 15% or the number of cell
types with a percentage above 30% exceeded two, we designated the spot
as unidentified. The inclusion of the “unidentified” category allows our
method to classify the ST spots with low cellular content or unknown
cell types. The dominant cell types were defined as those with the
highest proportions.</p>
</div>
<div id="installation" class="section level1">
<h1>2 Installation</h1>
<div id="prerequisites" class="section level2">
<h2>2.1 Prerequisites</h2>
<p>The following components were used to run the package. Different
versions of each may or may not be compatible.</p>
<ul>
<li>R (&gt;= 4.3.2)</li>
<li>GSVA</li>
<li>copykat: copykat[V1.1.0] (<a href="https://github.com/navinlabcode/copykat" class="uri">https://github.com/navinlabcode/copykat</a>)</li>
<li>dplyr</li>
</ul>
</div>
<div id="installing" class="section level2">
<h2>2.2 Installing</h2>
<p>To install the package, start R and enter:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">install_github</span>(<span class="st">&quot;wang-lab/SpatialDeX&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="running-spatialdex" class="section level1">
<h1>3 Running SpatialDeX</h1>
<p>Once SpatialDeX is installed, it can be loaded with the following
command:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(SpatialDeX)</span></code></pre></div>
<div id="running-deconvolution-example" class="section level2">
<h2>3.1 Running deconvolution example</h2>
<p>An illustrative example of a raw UMI matrix derived from a subset of
a tumor sample sequenced using the 10X Genomics platform is packaged
within SpatialDeX. This matrix, termed “example_raw_count”, each row
represents a distinct gene and each column represents a unique spot
barcode.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;example_raw_count&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>example_raw_count[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span></code></pre></div>
<table>
<colgroup>
<col width="7%" />
<col width="23%" />
<col width="23%" />
<col width="23%" />
<col width="23%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">AGTTATTGAAAGGTAA-1</th>
<th align="right">ATAAACCATTGGACGG-1</th>
<th align="right">ATAAACGGACCCGTAA-1</th>
<th align="right">ATAAAGGTCAAGTACG-1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">SP100</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">CAB39</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">ITM2C</td>
<td align="right">1</td>
<td align="right">10</td>
<td align="right">2</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">PSMD1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">NCL</td>
<td align="right">1</td>
<td align="right">7</td>
<td align="right">4</td>
<td align="right">3</td>
</tr>
</tbody>
</table>
<p>To test the package, run this line of code in R/Rstudio:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>res<span class="ot">&lt;-</span> SpatialDeX<span class="sc">::</span><span class="fu">SpatialDeX</span>(<span class="at">raw_exprs=</span>example_raw_count,</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>    <span class="at">lambda=</span><span class="dv">1</span>,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>    <span class="at">sam.name=</span><span class="st">&quot;test&quot;</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>    <span class="at">LOW.DR=</span><span class="fl">0.03</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>    <span class="at">UP.DR=</span><span class="fl">0.2</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>    <span class="at">ngene.chr=</span><span class="dv">5</span>,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>    <span class="at">KS.cut=</span><span class="fl">0.1</span>,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>    <span class="at">win.size=</span><span class="dv">25</span>,</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>    <span class="at">top_cell_type=</span><span class="dv">7</span>)</span></code></pre></div>
<div id="parameters" class="section level3">
<h3>3.1.1 Parameters</h3>
<p><strong><em>lambda</em></strong> lambda is the hyperparameter used
for regularization in ridge regression. Default is 1.</p>
<p><strong><em>sam.name</em></strong> is sample name.</p>
<p><strong><em>LOW.DR</em></strong> is minimal population fractions of
genes for smoothing for CNV detection. Default is 0.03.</p>
<p><strong><em>UP.DR</em></strong> is minimal population fractions of
genes for segmentation for CNV detection. Default is 0.2.</p>
<p><strong><em>ngene.chr</em></strong> is minimal number of genes per
chromosome for cell filtering for CNV detection. Default is 5.</p>
<p><strong><em>KS.cut</em></strong> is segmentation parameters, input 0
to 1; larger looser criteria for CNV detection. Default is 0.1.</p>
<p><strong><em>win.size</em></strong> is minimal window sizes for
segmentation for CNV detection. Default is 25.</p>
<p><strong><em>top_cell_type</em></strong> define the number of cell
types to retain based on the proportion of each cell type. Choose an
integer from 0 to 9.</p>
</div>
<div id="output" class="section level3">
<h3>3.1.2 Output</h3>
<p>Our analysis primarily targets major cell types, encompassing
malignant cells, T cells, B/plasma cells, macrophages, dendritic cells
(DCs), mast cells, endothelial cells, NK cells, and fibroblasts. The
resulting output comprises a dataframe containing spot barcodes and the
proportions of these major cell types. In this example, the parameter
“top_cell_type” is set to “7” to retain the cell types with the highest
seven proportions.</p>
<table>
<colgroup>
<col width="16%" />
<col width="8%" />
<col width="13%" />
<col width="8%" />
<col width="10%" />
<col width="8%" />
<col width="9%" />
<col width="10%" />
<col width="10%" />
<col width="2%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Malignant</th>
<th align="right">B.plasma_cells</th>
<th align="right">T_cells</th>
<th align="right">Endothelial</th>
<th align="right">DC</th>
<th align="right">Mast_cells</th>
<th align="right">Fibroblasts</th>
<th align="right">Macrophages</th>
<th align="right">NK</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AGTTATTGAAAGGTAA-1</td>
<td align="right">0.1206312</td>
<td align="right">0.1311543</td>
<td align="right">0.1527495</td>
<td align="right">0.0000000</td>
<td align="right">0.0702992</td>
<td align="right">0.1988094</td>
<td align="right">0.1842909</td>
<td align="right">0.1420654</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ATAAACCATTGGACGG-1</td>
<td align="right">0.1973387</td>
<td align="right">0.2149654</td>
<td align="right">0.1419585</td>
<td align="right">0.0876134</td>
<td align="right">0.0975152</td>
<td align="right">0.0000000</td>
<td align="right">0.1147252</td>
<td align="right">0.1458835</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ATAAACGGACCCGTAA-1</td>
<td align="right">0.2132504</td>
<td align="right">0.1211811</td>
<td align="right">0.1181572</td>
<td align="right">0.0000000</td>
<td align="right">0.1399636</td>
<td align="right">0.0988841</td>
<td align="right">0.1854656</td>
<td align="right">0.1230980</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ATAAAGGTCAAGTACG-1</td>
<td align="right">0.1877578</td>
<td align="right">0.1750442</td>
<td align="right">0.2445982</td>
<td align="right">0.0596204</td>
<td align="right">0.1587127</td>
<td align="right">0.1100996</td>
<td align="right">0.0641669</td>
<td align="right">0.0000000</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ATAACGTTACCTCCAC-1</td>
<td align="right">0.1420818</td>
<td align="right">0.1244093</td>
<td align="right">0.1858968</td>
<td align="right">0.0000000</td>
<td align="right">0.0754394</td>
<td align="right">0.0888839</td>
<td align="right">0.1634000</td>
<td align="right">0.2198887</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="get-the-weighted-feature-coefficient" class="section level2">
<h2>3.2 Get the weighted feature coefficient</h2>
<p>This function allows users to obtain their own weighted feature
coefficient matrix. Below provides an example of identity of cell types
across different spots. This identify matrix can be provided by users.
Each row represents a spot, and each column represents a cell type. A
value of 1 indicates the presence of that cell type in the spot, while 0
indicates its absence.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>raw_exprs <span class="ot">&lt;-</span> example_raw_count[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>cell_identity <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow =</span> <span class="dv">10</span>, <span class="at">ncol =</span> <span class="dv">9</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">rownames</span>(cell_identity) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(raw_exprs)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">colnames</span>(cell_identity) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;cell_type&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>)</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>) {</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  n_ones <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span>)</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="cf">if</span> (n_ones <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    ones_positions <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>, n_ones)</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>    cell_identity[i, ones_positions] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>  }</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>}</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>cell_identity[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<table>
<colgroup>
<col width="25%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">cell_type1</th>
<th align="right">cell_type2</th>
<th align="right">cell_type3</th>
<th align="right">cell_type4</th>
<th align="right">cell_type5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AGTTATTGAAAGGTAA-1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ATAAACCATTGGACGG-1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ATAAACGGACCCGTAA-1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ATAAAGGTCAAGTACG-1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ATAACGTTACCTCCAC-1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ATACAGCGTCCACTGA-1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ATACCCTGGCTCAAAT-1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ATACCGTCATCCATAA-1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">ATACGACAGATGGGTA-1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">ATACGCCGATCTACCG-1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>Run the code below to get the weighted feature coefficient
matrix.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>feat_coef_mat<span class="ot">&lt;-</span> <span class="fu">get_feat_coefficient</span>(cell_identity,</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                                <span class="at">raw_exprs=</span>raw_exprs,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                                <span class="at">LAMBDA=</span><span class="fl">0.01</span>,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>                                <span class="at">n_permutation=</span><span class="dv">10000</span>,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>                                <span class="at">pvalue=</span><span class="fl">0.05</span>,</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>                                <span class="at">sam.name=</span><span class="st">&quot;sam.name&quot;</span>,</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>                                <span class="at">LOW.DR=</span><span class="fl">0.03</span>,</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>                                <span class="at">UP.DR=</span><span class="fl">0.2</span>,</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>                                <span class="at">ngene.chr=</span><span class="dv">5</span>,</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>                                <span class="at">KS.cut=</span><span class="fl">0.1</span>,</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>                                <span class="at">win.size=</span><span class="dv">25</span>,</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>                                <span class="at">seed=</span><span class="dv">1234</span>)</span></code></pre></div>
<div id="parameters-1" class="section level3">
<h3>3.2.1 Parameters</h3>
<p><strong><em>n_permutation</em></strong> n_permutation An integer
indicating the number of permutations for randomly shuffling the cell
labels of the input matrix to create a null distribution. Default is
10,000.</p>
<p><strong><em>LAMBDA</em></strong> A numeric value for the
regularization parameter. Default is 0.01.</p>
<p><strong><em>sam.name</em></strong> Sample name</p>
<p><strong><em>LOW.DR</em></strong> Minimal population fractions of
genes for smoothing for CNV detection. Default is 0.03.</p>
<p><strong><em>UP.DR</em></strong> Minimal population fractions of genes
for segmentation for CNV detection. Default is 0.2.</p>
<p><strong><em>ngene.chr</em></strong> Minimal number of genes per
chromosome for cell filtering for CNV detection. Default is 5.</p>
<p><strong><em>KS.cut</em></strong> Segmentation parameters, input 0 to
1; larger looser criteria for CNV detection. Default is 0.1.</p>
<p><strong><em>win.size</em></strong> Minimal window sizes for
segmentation for CNV detection. Default is 25.</p>
<p><strong><em>pvalue</em></strong> Significance threshold for weighted
feature coefficients. Calculated by permutation.</p>
<p><strong><em>seed</em></strong> Seed for random number generation for
reproducibility.</p>
</div>
<div id="output-1" class="section level3">
<h3>3.2.2 Output</h3>
<p>A matrix of coefficients for each feature to each cell type is
generated. Each cell in the matrix contains a coefficient indicating the
strength of the relationship between a feature and a given cell
type.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>feat_coef_mat[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<table>
<colgroup>
<col width="11%" />
<col width="21%" />
<col width="12%" />
<col width="18%" />
<col width="22%" />
<col width="14%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">htca_Epithelial_Cell</th>
<th align="right">htca_Muscle</th>
<th align="right">htca_Stromal_Cell</th>
<th align="right">htca_Endothelial_Cell</th>
<th align="right">htca_Monocyte</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">cell_type1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cell_type2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cell_type3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cell_type4</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1.588904</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cell_type5</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cell_type6</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cell_type7</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cell_type8</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cell_type9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0.000000</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="sessioninfo" class="section level1">
<h1>sessionInfo</h1>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">3.2</span> (<span class="dv">2023-10-31</span> ucrt)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>Platform<span class="sc">:</span> x86_64<span class="sc">-</span>w64<span class="sc">-</span>mingw32<span class="sc">/</span><span class="fu">x64</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>Running under<span class="sc">:</span> Windows <span class="dv">11</span> <span class="fu">x64</span> (build <span class="dv">22631</span>)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>[<span class="dv">1</span>] LC_COLLATE<span class="ot">=</span>English_United States.utf8  LC_CTYPE<span class="ot">=</span>English_United States.utf8    LC_MONETARY<span class="ot">=</span>English_United States.utf8</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>[<span class="dv">4</span>] LC_NUMERIC<span class="ot">=</span>C                           LC_TIME<span class="ot">=</span>English_United States.utf8    </span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>time zone<span class="sc">:</span> America<span class="sc">/</span>Chicago</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>tzcode source<span class="sc">:</span> internal</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>[<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>[<span class="dv">1</span>] SpatialDeX_2.<span class="fl">99.0</span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a>  [<span class="dv">1</span>] DBI_1.<span class="fl">2.2</span>                   bitops_1<span class="fl">.0</span><span class="dv">-7</span>                GSEABase_1.<span class="fl">64.0</span>             remotes_2.<span class="dv">4</span>.<span class="fl">2.1</span>            </span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a>  [<span class="dv">5</span>] testthat_3.<span class="fl">2.1</span>              rlang_1.<span class="fl">1.3</span>                 magrittr_2.<span class="fl">0.3</span>              matrixStats_1.<span class="fl">2.0</span>          </span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a>  [<span class="dv">9</span>] compiler_4.<span class="fl">3.2</span>              RSQLite_2.<span class="fl">3.5</span>               DelayedMatrixStats_1.<span class="fl">24.0</span>   roxygen2_7.<span class="fl">3.1</span>             </span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a> [<span class="dv">13</span>] callr_3.<span class="fl">7.5</span>                 png_0<span class="fl">.1</span><span class="dv">-7</span>                   vctrs_0.<span class="fl">6.5</span>                 stringr_1.<span class="fl">5.1</span>              </span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a> [<span class="dv">17</span>] profvis_0.<span class="fl">3.8</span>               pkgconfig_2.<span class="fl">0.3</span>             crayon_1.<span class="fl">5.2</span>                fastmap_1.<span class="fl">1.1</span>              </span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a> [<span class="dv">21</span>] XVector_0.<span class="fl">42.0</span>              ellipsis_0.<span class="fl">3.2</span>              utf8_1.<span class="fl">2.2</span>                  promises_1.<span class="dv">2</span>.<span class="fl">0.1</span>           </span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a> [<span class="dv">25</span>] rmarkdown_2<span class="fl">.26</span>              sessioninfo_1.<span class="fl">2.2</span>           ps_1.<span class="fl">7.6</span>                    graph_1.<span class="fl">80.0</span>               </span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a> [<span class="dv">29</span>] purrr_1.<span class="fl">0.2</span>                 bit_4.<span class="fl">0.4</span>                   xfun_0<span class="fl">.42</span>                   beachmat_2.<span class="fl">18.1</span>            </span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a> [<span class="dv">33</span>] zlibbioc_1.<span class="fl">48.0</span>             cachem_1.<span class="fl">0.8</span>                GenomeInfoDb_1.<span class="fl">38.6</span>         blob_1.<span class="fl">2.4</span>                 </span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a> [<span class="dv">37</span>] later_1.<span class="fl">3.0</span>                 rhdf5filters_1.<span class="fl">14.1</span>         DelayedArray_0.<span class="fl">28.0</span>         Rhdf5lib_1.<span class="fl">24.2</span>            </span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a> [<span class="dv">41</span>] BiocParallel_1.<span class="fl">36.0</span>         prettyunits_1.<span class="fl">2.0</span>           irlba_2.<span class="fl">3.3</span>                 parallel_4.<span class="fl">3.2</span>             </span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a> [<span class="dv">45</span>] R6_2.<span class="fl">5.1</span>                    stringi_1.<span class="fl">7.5</span>               pkgload_1.<span class="fl">3.4</span>               brio_1.<span class="fl">1.4</span>                 </span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a> [<span class="dv">49</span>] GenomicRanges_1.<span class="fl">54.1</span>        Rcpp_1.<span class="fl">0.8</span>                  SummarizedExperiment_1.<span class="fl">32.0</span> knitr_1<span class="fl">.45</span>                 </span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a> [<span class="dv">53</span>] copykat_1.<span class="fl">0.8</span>               usethis_2.<span class="fl">1.6</span>               GSVA_1.<span class="fl">50.0</span>                 IRanges_2.<span class="fl">36.0</span>             </span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a> [<span class="dv">57</span>] httpuv_1.<span class="fl">6.3</span>                Matrix_1<span class="fl">.6</span><span class="dv">-5</span>                tidyselect_1.<span class="fl">2.0</span>            yaml_2.<span class="fl">3.8</span>                 </span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a> [<span class="dv">61</span>] rstudioapi_0.<span class="fl">15.0</span>           abind_1<span class="fl">.4</span><span class="dv">-5</span>                 codetools_0<span class="fl">.2</span><span class="dv">-19</span>            miniUI_0.<span class="dv">1</span>.<span class="fl">1.1</span>             </span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a> [<span class="dv">65</span>] curl_5.<span class="fl">2.0</span>                  processx_3.<span class="fl">8.3</span>              pkgbuild_1.<span class="fl">4.3</span>              lattice_0<span class="fl">.22</span><span class="dv">-5</span>             </span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a> [<span class="dv">69</span>] tibble_3.<span class="fl">1.6</span>                Biobase_2.<span class="fl">62.0</span>              shiny_1.<span class="fl">8.0</span>                 withr_3.<span class="fl">0.0</span>                </span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a> [<span class="dv">73</span>] KEGGREST_1.<span class="fl">42.0</span>             evaluate_0<span class="fl">.23</span>               desc_1.<span class="fl">4.3</span>                  urlchecker_1.<span class="fl">0.1</span>           </span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a> [<span class="dv">77</span>] xml2_1.<span class="fl">3.3</span>                  Biostrings_2.<span class="fl">70.2</span>           pillar_1.<span class="fl">9.0</span>                rsconnect_1.<span class="fl">2.1</span>            </span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a> [<span class="dv">81</span>] MatrixGenerics_1.<span class="fl">14.0</span>       stats4_4.<span class="fl">3.2</span>                generics_0.<span class="fl">1.3</span>              xopen_1.<span class="fl">0.0</span>                </span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a> [<span class="dv">85</span>] rprojroot_2.<span class="fl">0.4</span>             RCurl_1<span class="fl">.98-1.14</span>             S4Vectors_0.<span class="fl">40.2</span>            sparseMatrixStats_1.<span class="fl">14.0</span>   </span>
<span id="cb10-44"><a href="#cb10-44" tabindex="-1"></a> [<span class="dv">89</span>] xtable_1<span class="fl">.8</span><span class="dv">-4</span>                glue_1.<span class="fl">6.2</span>                  tools_4.<span class="fl">3.2</span>                 ScaledMatrix_1.<span class="fl">10.0</span>        </span>
<span id="cb10-45"><a href="#cb10-45" tabindex="-1"></a> [<span class="dv">93</span>] annotate_1.<span class="fl">80.0</span>             fs_1.<span class="fl">5.2</span>                    XML_3<span class="fl">.99</span><span class="dv">-0</span>.<span class="fl">16.1</span>             rhdf5_2.<span class="fl">46.1</span>               </span>
<span id="cb10-46"><a href="#cb10-46" tabindex="-1"></a> [<span class="dv">97</span>] grid_4.<span class="fl">3.2</span>                  devtools_2.<span class="fl">4.5</span>              AnnotationDbi_1.<span class="fl">64.1</span>        SingleCellExperiment_1.<span class="fl">24.0</span></span>
<span id="cb10-47"><a href="#cb10-47" tabindex="-1"></a>[<span class="dv">101</span>] GenomeInfoDbData_1.<span class="fl">2.11</span>     BiocSingular_1.<span class="fl">18.0</span>         HDF5Array_1.<span class="fl">30.1</span>            rsvd_1.<span class="fl">0.5</span>                 </span>
<span id="cb10-48"><a href="#cb10-48" tabindex="-1"></a>[<span class="dv">105</span>] cli_3.<span class="fl">6.2</span>                   rcmdcheck_1.<span class="fl">4.0</span>             fansi_1.<span class="fl">0.2</span>                 S4Arrays_1.<span class="fl">2.0</span>             </span>
<span id="cb10-49"><a href="#cb10-49" tabindex="-1"></a>[<span class="dv">109</span>] dplyr_1.<span class="fl">0.7</span>                 digest_0.<span class="fl">6.29</span>               BiocGenerics_0.<span class="fl">48.1</span>         SparseArray_1.<span class="fl">2.4</span>          </span>
<span id="cb10-50"><a href="#cb10-50" tabindex="-1"></a>[<span class="dv">113</span>] htmlwidgets_1.<span class="fl">6.4</span>           memoise_2.<span class="fl">0.1</span>               htmltools_0.<span class="fl">5.7</span>             lifecycle_1.<span class="fl">0.4</span>            </span>
<span id="cb10-51"><a href="#cb10-51" tabindex="-1"></a>[<span class="dv">117</span>] httr_1.<span class="fl">4.7</span>                  mime_0<span class="fl">.12</span>                   bit64_4.<span class="fl">0.5</span>               </span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
